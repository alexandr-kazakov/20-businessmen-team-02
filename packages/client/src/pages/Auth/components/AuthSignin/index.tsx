import React, { useEffect, useState } from 'react'
import { useHistory } from 'react-router-dom'
import { Controller, type FieldValues, useForm } from 'react-hook-form'

import { useAppDispatch } from '../../../../app/redux/hooks'
import { signin, setIsSigninView, getOAuthUrl } from '../../redux/authSlice'
import { showSnackBar } from '../../../../components/Snackbar/redux/snackbarSlice'
import { Input } from '../../../../components/UI/Input'
import { Button, ButtonVariant } from '../../../../components/UI/Button'
import { RoutersPaths } from '../../../../components/Routers/types'
import { LOGIN_REGEXP, PASSWORD_REGEXP } from '../../../../lib/regexp'

import styles from './styles.module.scss'

const INIT_VALUES = {
  login: '',
  password: '',
}

export const AuthSignIn: React.FC = () => {
  const history = useHistory()
  const dispatch = useAppDispatch()

  const [oAuthYandexUrl, setOAuthYandexUrl] = useState('')

  const { control, handleSubmit } = useForm({ defaultValues: INIT_VALUES })

  const onSubmit = async (data: FieldValues): Promise<void> => {
    const response = await dispatch(signin(data))

    if (response.error) {
      dispatch(showSnackBar(response.error.message))
    } else {
      history.push(RoutersPaths.main)
    }
  }

  const handlerToggle = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault()
    dispatch(setIsSigninView())
  }

  useEffect(() => {
    if (!oAuthYandexUrl) {
      getOAuthUrl().then(setOAuthYandexUrl)
    }
  }, [oAuthYandexUrl])

  return (
    <form className={styles.form} onSubmit={handleSubmit(onSubmit)}>
      <span className={styles.title}>Вход</span>
      <div className={styles.inputs}>
        <Controller
          name="login"
          control={control}
          rules={{ required: true, pattern: LOGIN_REGEXP }}
          render={({ field, fieldState }) => (
            <Input
              value={field.value}
              onChange={field.onChange}
              placeholder="Логин"
              isValid={!(fieldState.error && 'error')}
            />
          )}
        />

        <Controller
          name="password"
          control={control}
          rules={{ required: true, pattern: PASSWORD_REGEXP }}
          render={({ field, fieldState }) => (
            <Input
              type="password"
              value={field.value}
              onChange={field.onChange}
              placeholder="Пароль"
              isValid={!(fieldState.error && 'error')}
            />
          )}
        />
      </div>
      <div className={styles.buttons}>
        <Button type="submit">Войти</Button>
        <a href={oAuthYandexUrl} className={styles.yandexButton}>
          <Button type="button">Войти через YandexId</Button>
        </a>
        <Button onClick={handlerToggle} variant={ButtonVariant.SECONDARY}>
          Нет аккаунта?
        </Button>
      </div>
    </form>
  )
}
