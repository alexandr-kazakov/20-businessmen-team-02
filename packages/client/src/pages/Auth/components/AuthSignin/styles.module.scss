@use '../../../../styles/common.module.scss';

.form {
  display: flex;
  flex-direction: column;
  max-width: 340px;
  width: 100%;
  border: 1px solid #eee;
  border-radius: 12px;
  padding: 30px;
  box-shadow: 0px 0px 6px rgba(0, 0, 0, 0.14);
}

.title {
  text-align: center;
  margin-bottom: 20px;
  font-style: normal;
  font-weight: 500;
  font-size: 20px;
  line-height: 20px;
  color: #fff;
}

.inputs {
  display: flex;
  flex-direction: column;
  gap: 10px;
  margin-bottom: 20px;
}

.yandexButton {
  display: block;
  width: 100%;

  & button {
    background-color: #fc0 !important;
    color: #000;
    text-decoration: none;
  }
}
