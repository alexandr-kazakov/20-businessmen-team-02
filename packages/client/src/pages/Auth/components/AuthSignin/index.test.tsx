import React from 'react'
import { render } from '@testing-library/react'
import { AuthSignIn } from '@/pages/Auth/components/AuthSignin/index'

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn()
}));

test('should render AuthSignIn component', () => {
  const {asFragment} = render(<AuthSignIn/>)

  expect(asFragment()).toMatchSnapshot()
})