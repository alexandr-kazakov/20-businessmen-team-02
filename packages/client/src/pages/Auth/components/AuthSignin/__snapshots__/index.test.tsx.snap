// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AuthSignIn component 1`] = `
<DocumentFragment>
  <form>
    <span>
      Вход
    </span>
    <div>
      <input
        class=""
        placeholder="Логин"
        value=""
      />
      <input
        class=""
        placeholder="Пароль"
        type="password"
        value=""
      />
    </div>
    <div>
      <button
        class=""
        type="submit"
      >
        Войти
      </button>
      <a
        href=""
      >
        <button
          class=""
          type="button"
        >
          Войти через YandexId
        </button>
      </a>
      <button
        class="undefined"
      >
        Нет аккаунта?
      </button>
    </div>
  </form>
</DocumentFragment>
`;
